src/ophyd_async/epics/areadetector/writers/general_hdffile.py

```diff
@@ -1,11 +1,31 @@
+from dataclasses import dataclass
 from pathlib import Path
-from typing import Iterator, List
+from typing import Any, Iterator, List, Optional
 
-from event_model import StreamDatum, StreamResource, compose_stream_resource
+from event_model import (
+    ComposeStreamResource,
+    ComposeStreamResourceBundle,
+    StreamDatum,
+    StreamResource,
+)
 
 from ophyd_async.core import DirectoryInfo
 
-from ._hdfdataset import _HDFDataset
+
+@dataclass
+class _HDFDataset:
```

Comment:
Per @coretl:

This should be slimmed to only what's needed to describe a file, not what's needed to _write_ it.

The pattern detector is different from the others because it is actually _writing_ the HDF5 file, not describing a file written by an external program (IOC). For the pattern detector, separate dataclass should _extend_ this to add `maxshape` and `fillvalue`.

With that separtion, we can make all of these fields required fields.

---

Some of this information is used to construct the EventDescriptor (`shape`, `dtype`, `name` ?, `multiplier`). Other information is used to construct the StreamResource (`path`, meaning the "path" of the dataset inside the layout of the HDF5 file, distinct from the file; and optionally `smwr`).

Add `dtype_str`, the numpy-style dtype like `"<i4"` or `"<f8"`.

This should also include the `uri`. (For old-style StreamResource we can convert this into `root` and `resource_path` by ignoring the hostname, hard-coding `root` to `"/"` as discussed, and using the rest of the path as `resource_path`.)

https://github.com/bluesky/bluesky/blob/bc8222be2e099a00baafcede1d761d61b213bf19/src/bluesky/callbacks/tiled_writer.py#L167-L178